In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

from numpy import mean
from numpy import std
from matplotlib import pyplot
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report
from sklearn.preprocessing import QuantileTransformer
from sklearn.pipeline import Pipeline

In [ ]:
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

In [ ]:
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

In [ ]:
!pip install mne

     |████████████████████████████████| 7.4 MB 7.2 MB/s 


In [ ]:
#import pyeeg
from tensorflow import keras 

In [ ]:
%matplotlib inline

import glob
import math
import ntpath
import random, os, sys
import numpy as np
import pandas as pd
import tensorflow as tf
import matplotlib.pyplot as plt
import seaborn as sns
#import edfreader
#import models
import mne
#import pyeeg
import warnings
warnings.filterwarnings('ignore')

from matplotlib.backends.backend_pdf import PdfPages
from mne.datasets.sleep_physionet._utils import _fetch_one, _data_path, AGE_SLEEP_RECORDS, _check_subjects
from datetime import datetime
from mne import Epochs, pick_types, find_events
from mne.io import concatenate_raws, read_raw_edf
from mne.time_frequency import psd_welch
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report

from tqdm.notebook import tqdm
from sklearn.model_selection import train_test_split
from tensorflow import keras

from tensorflow.keras.utils import to_categorical

from tensorflow.keras.callbacks import ModelCheckpoint, ReduceLROnPlateau, CSVLogger
import imblearn
from imblearn.over_sampling import SMOTE
from sklearn.svm import SVC
import joblib

from sklearn.metrics import make_scorer, f1_score, accuracy_score, classification_report, log_loss
from sklearn.metrics import roc_auc_score, confusion_matrix, roc_auc_score, roc_curve

from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier, GradientBoostingClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.discriminant_analysis import QuadraticDiscriminantAnalysis
#from xgboost import XGBClassifier
#import xgboost

In [ ]:
VBS = True  # constant boolean to enable/disbale verbose
EPOCH_SEC_SIZE = 30  # Epoch duration selection
seed = 42  # seed value for the random seeds
batch_size = 64
number_of_files = 100 #100 npz files, 50 subjects, each suject has 2 nights, so 2 files per subject (50*2)=100

# values to label the stages
UNKNOWN = -1
W = 0
N1 = 1
N2 = 2
N3 = 3
REM = 4

# making string dictionary for the label values
label_dict = {
    "UNKNOWN"  : UNKNOWN,
    "W"        : W,
    "N1"       : N1,
    "N2"       : N2,
    "N3"       : N3,
    "REM"      : REM
}

# converting from label values to strings 
class_dict = {
    -1: "UNKNOWN",
    0 : "W",
    1 : "N1",
    2 : "N2",
    3 : "N3",
    4 : "REM"
}

# annotation dictionary to convert from string to label values
annot2label = {
    "Sleep stage ?": -1,
    "Movement time": -1,
    "Sleep stage W": 0,
    "Sleep stage 1": 1,
    "Sleep stage 2": 2,
    "Sleep stage 3": 3,
    "Sleep stage 4": 3,
    "Sleep stage R": 4
}
#project_path = os.path.abspath(os.getcwd())  # finding the current project path in windows

In [ ]:
# localized fetch_data function by using mne library
# https://github.com/mne-tools/mne-python/blob/maint/0.20/mne/datasets/sleep_physionet/age.py#L18-L111
data_path = _data_path
BASE_URL = 'https://physionet.org/physiobank/database/sleep-edfx/sleep-cassette/'
def fetch_data(subjects, recording=[1, 2], path=None, force_update=False,
               update_path=None, base_url=BASE_URL,
               verbose=None):  # noqa: D301
    records = np.loadtxt(AGE_SLEEP_RECORDS,
                         skiprows=1,
                         delimiter=',',
                         usecols=(0, 1, 2, 6, 7),
                         dtype={'names': ('subject', 'record', 'type', 'sha',
                                          'fname'),
                                'formats': ('<i2', 'i1', '<S9', 'S40', '<S22')}
                         )
    psg_records = records[np.where(records['type'] == b'PSG')]
    hyp_records = records[np.where(records['type'] == b'Hypnogram')]

    path = data_path(path=path, update_path=update_path)
    params = [path, force_update, base_url]
    fnames = []
    for subject in subjects:
        for idx in np.where(psg_records['subject'] == subject)[0]:
            if psg_records['record'][idx] in recording:
                psg_fname = _fetch_one(psg_records['fname'][idx].decode(),
                                       psg_records['sha'][idx].decode(),
                                       *params)
                hyp_fname = _fetch_one(hyp_records['fname'][idx].decode(),
                                       hyp_records['sha'][idx].decode(),
                                       *params)
                fnames.append([psg_fname, hyp_fname])

    return fnames

In [ ]:
from google.colab import drive
drive.mount('/gdrive')
%cd /gdrive

Mounted at /gdrive
/gdrive


In [ ]:
npz_files = sorted(glob.glob("/gdrive/My Drive/NPZ_files/***.npz"))
X = np.zeros((0, 3000, 1))
y = []
for fn in tqdm(npz_files[:number_of_files]):
    samples = np.load(fn)
    X_data = samples['x']
    X = np.concatenate((X, X_data), axis=0)
    y.extend(samples['y'])
y = np.array(y)

#sorted(glob.glob("/gdrive/My Drive/NPZ_files/***.npz"))


  0%|          | 0/100 [00:00<?, ?it/s]

In [ ]:
npz_files

['/gdrive/My Drive/NPZ_files/SC4001E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4002E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4011E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4012E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4021E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4022E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4031E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4032E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4041E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4042E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4051E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4052E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4061E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4062E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4071E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4072E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4081E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4082E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4091E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4092E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4101E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4102E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4

In [ ]:
len(npz_files[:number_of_files])

100

In [ ]:
npz_files[:number_of_files]

['/gdrive/My Drive/NPZ_files/SC4001E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4002E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4011E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4012E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4021E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4022E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4031E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4032E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4041E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4042E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4051E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4052E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4061E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4062E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4071E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4072E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4081E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4082E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4091E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4092E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4101E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4102E0.npz',
 '/gdrive/My Drive/NPZ_files/SC4

In [ ]:
npz_files[148:150]

['/gdrive/My Drive/NPZ_files/SC4821G0.npz',
 '/gdrive/My Drive/NPZ_files/SC4822G0.npz']

In [ ]:
X

array([[[ 8.11135578e+00],
        [ 1.74886456e+01],
        [ 2.12395611e+01],
        ...,
        [-1.03619051e+01],
        [-1.11120882e+01],
        [-2.10989022e+00]],

       [[-1.07369967e+01],
        [-1.13934069e+01],
        [-4.45421267e+00],
        ...,
        [ 5.88424911e+01],
        [ 4.83399277e+01],
        [ 5.36849823e+01]],

       [[ 6.13743591e+01],
        [ 3.86813202e+01],
        [ 4.95589752e+01],
        ...,
        [ 3.36175842e+01],
        [ 3.43677673e+01],
        [ 3.18358974e+01]],

       ...,

       [[ 3.47106218e+00],
        [-5.03052510e-02],
        [ 2.56556773e+00],
        ...,
        [ 9.55799758e-01],
        [-1.86129427e+00],
        [ 3.37045169e+00]],

       [[ 2.26373625e+00],
        [ 3.52136761e-01],
        [ 2.86739922e+00],
        ...,
        [ 2.54041519e+01],
        [ 2.02730160e+01],
        [ 1.82608051e+01]],

       [[ 1.96693535e+01],
        [ 1.65504265e+01],
        [ 1.54437122e+01],
        ...,
        

In [ ]:
y

array([0, 0, 0, ..., 0, 0, 0], dtype=int32)

In [ ]:
from keras.utils import np_utils

In [ ]:
if VBS:
    print("Shape of the input data: {}".format(X.shape))
    print("Shape of the sleep stages: {}".format(y.shape))


Shape of the input data: (116718, 3000, 1)
Shape of the sleep stages: (116718,)


In [ ]:
from tensorflow.keras import optimizers, losses, activations, models
from tensorflow.keras.models import Model, load_model
from tensorflow.keras.layers import GlobalMaxPool1D, GlobalAveragePooling1D, concatenate, SpatialDropout1D, TimeDistributed, Bidirectional, LSTM
from tensorflow.keras.layers import Input, Conv1D, Dense, Dropout, MaxPool1D, Activation
from tensorflow.keras.layers import Reshape, LSTM, TimeDistributed, Bidirectional, BatchNormalization, Flatten, RepeatVector
from tensorflow.keras.optimizers import Adam
#from keras_contrib.layers import CRF

from scipy.signal import butter, lfilter


In [ ]:
#n_classes=5
def model_b(n_classes=5, use_sub_layer=False, use_rnn=True, verbose=False):
    inputLayer = Input(shape=(3000, 1), name='inLayer')
    convFine = Conv1D(filters=64, kernel_size=int(Fs/2), strides=int(Fs/16), padding='same', activation='relu', name='fConv1')(inputLayer)
    convFine = MaxPool1D(pool_size=8, strides=8, name='fMaxP1')(convFine)
    convFine = Dropout(rate=0.5, name='fDrop1')(convFine)
    convFine = Conv1D(filters=128, kernel_size=8, padding='same', activation='relu', name='fConv2')(convFine)
    convFine = Conv1D(filters=128, kernel_size=8, padding='same', activation='relu', name='fConv3')(convFine)
    convFine = Conv1D(filters=128, kernel_size=8, padding='same', activation='relu', name='fConv4')(convFine)
    convFine = MaxPool1D(pool_size=4, strides=4, name='fMaxP2')(convFine)
    fineShape = convFine.get_shape()
    convFine = Flatten(name='fFlat1')(convFine)
    
    # network to learn coarse features
    convCoarse = Conv1D(filters=32, kernel_size=Fs*4, strides=int(Fs/2), padding='same', activation='relu', name='cConv1')(inputLayer)
    convCoarse = MaxPool1D(pool_size=4, strides=4, name='cMaxP1')(convCoarse)
    convCoarse = Dropout(rate=0.5, name='cDrop1')(convCoarse)
    convCoarse = Conv1D(filters=128, kernel_size=6, padding='same', activation='relu', name='cConv2')(convCoarse)
    convCoarse = Conv1D(filters=128, kernel_size=6, padding='same', activation='relu', name='cConv3')(convCoarse)
    convCoarse = Conv1D(filters=128, kernel_size=6, padding='same', activation='relu', name='cConv4')(convCoarse)
    convCoarse = MaxPool1D(pool_size=2, strides=2, name='cMaxP2')(convCoarse)
    coarseShape = convCoarse.get_shape()
    convCoarse = Flatten(name='cFlat1')(convCoarse)
    
    # concatenate coarse and fine cnns
    mergeLayer = concatenate([convFine, convCoarse], name='merge_1')
    outLayer = Dropout(rate=0.5, name='mDrop1')(mergeLayer)
    
    outLayer = Reshape((1, outLayer.get_shape()[1]), name='reshape1')(outLayer)
    outLayer = LSTM(64, return_sequences=True)(outLayer)
    outLayer = LSTM(64, return_sequences=False)(outLayer)

    # Classify
    outLayer = Dense(n_classes, activation='softmax', name='outLayer')(outLayer)
    model = Model(inputLayer, outLayer)
    optimizer = Adam(lr=1e-4)
    model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['acc'])
    #model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])
    if verbose:
        model.summary()
    return model

In [ ]:
def butter_bandpass(lowcut, highpass, fs, order=4):
    nyq = 0.5 * fs
    #       low = lowcut / nyq
    high = highpass / nyq
    b, a = butter(order, high, btype='highpass')
    return b, a

In [ ]:
def butter_bandpass_filter(data, highpass, fs, order=4):
    b, a = butter_bandpass(0, highpass, fs, order=order)
    y = lfilter(b, a, data)
    return y

In [ ]:
#npz_files = sorted(glob.glob(os.path.join(output_path, "*.npz")))
X_sub_482 = np.zeros((0, 3000, 1))
y_sub_482 = []
for fn in tqdm(npz_files[148:150]): #extracting data from the npz files corresponding to both tge recording 1, 2 of the 50th subject (no 53rd subject)
    samples = np.load(fn)
    X_data = samples['x']
    X_sub_482 = np.concatenate((X_sub_482, X_data), axis=0)
    y_sub_482.extend(samples['y'])
y_sub_482 = np.array(y_sub_482)

  0%|          | 0/2 [00:00<?, ?it/s]

In [ ]:
X_sub_482

array([[[-49.10647202],
        [-48.57606888],
        [-45.21685028],
        ...,
        [  1.81220996],
        [ -2.51941395],
        [  3.75702071]],

       [[ 17.63589668],
        [  4.37582397],
        [ 25.59194183],
        ...,
        [ 52.20048904],
        [ 55.55970764],
        [ 57.06251526]],

       [[ 50.25567627],
        [ 55.64810562],
        [ 50.16727829],
        ...,
        [-13.03907204],
        [-16.75189209],
        [-13.12747288]],

       ...,

       [[  4.04444456],
        [  2.26666665],
        [  6.44444466],
        ...,
        [ -6.17777777],
        [ 10.17777824],
        [-48.93333435]],

       [[-43.95555496],
        [-63.42222214],
        [-70.8888855 ],
        ...,
        [ 20.66666603],
        [ 14.88888931],
        [ 22.35555649]],

       [[ 21.73333359],
        [ 21.55555534],
        [ 21.91111183],
        ...,
        [  1.28888893],
        [  2.97777772],
        [  5.28888893]]])

In [ ]:
y_sub_482

array([0, 0, 0, ..., 0, 0, 0], dtype=int32)

In [ ]:
if VBS:
    print("Shape of the subject(482) data: {}".format(X_sub_482.shape))
    print("Shape of the corresponding sleep stage: {}".format(y_sub_482.shape))

Shape of the subject(482) data: (3070, 3000, 1)
Shape of the corresponding sleep stage: (3070,)


In [ ]:
pp_x = np.array([butter_bandpass_filter(sample, highpass=40.0, fs=100, order=4) for sample in X])

if VBS:
    print(pp_x.shape)

(116718, 3000, 1)


In [ ]:
y.shape

(116718,)

In [ ]:
type(y)

numpy.ndarray

In [ ]:
pp_x

array([[[ 3.91319654e-02],
        [ 8.43712310e-02],
        [ 1.02466935e-01],
        ...,
        [-4.99893880e-02],
        [-5.36085289e-02],
        [-1.01788349e-02]],

       [[-5.17989585e-02],
        [-5.49657067e-02],
        [-2.14886513e-02],
        ...,
        [ 2.83876381e-01],
        [ 2.33208409e-01],
        [ 2.58994788e-01]],

       [[ 2.96090982e-01],
        [ 1.86611970e-01],
        [ 2.39089513e-01],
        ...,
        [ 1.62182769e-01],
        [ 1.65801910e-01],
        [ 1.53587300e-01]],

       ...,

       [[ 1.67455958e-02],
        [-2.42689804e-04],
        [ 1.23771796e-02],
        ...,
        [ 4.61110622e-03],
        [-8.97952265e-03],
        [ 1.62602162e-02]],

       [[ 1.09210409e-02],
        [ 1.69882864e-03],
        [ 1.38333184e-02],
        ...,
        [ 1.22558352e-01],
        [ 9.78039900e-02],
        [ 8.80963939e-02]],

       [[ 9.48917148e-02],
        [ 7.98449401e-02],
        [ 7.45057705e-02],
        ...,
        

In [ ]:
pp_x_rshp=pp_x.reshape(116718, 3000)

In [ ]:
pp_x_rshp_df=pd.DataFrame(pp_x_rshp)

In [ ]:
pp_x_rshp_df['sleep_stage']=y

In [ ]:
pp_x_rshp_df

0         1         2  ...      2998      2999  sleep_stage
0       0.039132  0.084371  0.102467  ... -0.053609 -0.010179            0
1      -0.051799 -0.054966 -0.021489  ...  0.233208  0.258995            0
2       0.296091  0.186612  0.239090  ...  0.165802  0.153587            0
3       0.134134  0.116491  0.116491  ...  0.205612  0.274829            0
4       0.333640  0.359878  0.381141  ...  0.122372  0.207422            0
...          ...       ...       ...  ...       ...       ...          ...
116713  0.117219  0.233710  0.088582  ... -0.099746 -0.122073            0
116714 -0.087611 -0.076447 -0.056061  ...  0.046839  0.049266            0
116715  0.016746 -0.000243  0.012377  ... -0.008980  0.016260            0
116716  0.010921  0.001699  0.013833  ...  0.097804  0.088096            0
116717  0.094892  0.079845  0.074506  ...  0.143430  0.085184            0

[116718 rows x 3001 columns]

In [ ]:
dt_shuf2=pp_x_rshp_df.sample(frac=1,replace=True, random_state=456)

In [ ]:
dt_shuf2

0         1         2  ...      2998      2999  sleep_stage
30619  0.028727  0.019679 -0.133682  ... -0.050894 -0.031441            0
66149 -0.064884 -0.093265 -0.132826  ... -0.026183 -0.034784            2
48811 -0.059500  0.146904 -0.240929  ...  0.134180  0.155858            0
92761 -0.000707 -0.031338 -0.033694  ... -0.019085 -0.030866            2
55338  0.041534  0.042803  0.060566  ... -0.009642  0.008545            2
...         ...       ...       ...  ...       ...       ...          ...
51593  0.031014  0.023804  0.028469  ... -0.033452 -0.022849            2
18057 -0.384249 -0.389748 -0.375541  ... -0.275636 -0.285259            3
85379 -0.008700 -0.039090 -0.050882  ... -0.005525 -0.000990            0
26462  0.011455  0.039213  0.040054  ... -0.046586 -0.033127            4
68917  0.024599  0.027144  0.011875  ...  0.047077  0.064042            2

[116718 rows x 3001 columns]

In [ ]:
dt_shuf2['sleep_stage']

30619    0
66149    2
48811    0
92761    2
55338    2
        ..
51593    2
18057    3
85379    0
26462    4
68917    2
Name: sleep_stage, Length: 116718, dtype: int32

In [ ]:
x_shuf2=dt_shuf2.drop('sleep_stage',axis=1)
y_shuf2=dt_shuf2['sleep_stage']

In [ ]:
x_shuf2

0         1         2     ...      2997      2998      2999
30619  0.028727  0.019679 -0.133682  ...  0.112420 -0.050894 -0.031441
66149 -0.064884 -0.093265 -0.132826  ... -0.018443 -0.026183 -0.034784
48811 -0.059500  0.146904 -0.240929  ...  0.162455  0.134180  0.155858
92761 -0.000707 -0.031338 -0.033694  ... -0.025211 -0.019085 -0.030866
55338  0.041534  0.042803  0.060566  ... -0.004143 -0.009642  0.008545
...         ...       ...       ...  ...       ...       ...       ...
51593  0.031014  0.023804  0.028469  ... -0.032604 -0.033452 -0.022849
18057 -0.384249 -0.389748 -0.375541  ... -0.245389 -0.275636 -0.285259
85379 -0.008700 -0.039090 -0.050882  ...  0.008989 -0.005525 -0.000990
26462  0.011455  0.039213  0.040054  ... -0.066774 -0.046586 -0.033127
68917  0.024599  0.027144  0.011875  ...  0.035626  0.047077  0.064042

[116718 rows x 3000 columns]

In [ ]:
y_shuf2_=np_utils.to_categorical(y_shuf2)

In [ ]:
x_shuf2_rshp=np.array(x_shuf2).reshape(116718,3000,1)

In [ ]:
pp_X_sub_482 = np.array([butter_bandpass_filter(sample, highpass=40.0, fs=100, order=4) for sample in X_sub_482])

if VBS:
    print(pp_X_sub_482.shape)

(3070, 3000, 1)


In [ ]:
pp_X_sub_482

array([[[-0.23690648],
        [-0.23434764],
        [-0.21814161],
        ...,
        [ 0.00874272],
        [-0.01215452],
        [ 0.01812516]],

       [[ 0.08508162],
        [ 0.02111048],
        [ 0.12346431],
        ...,
        [ 0.25183308],
        [ 0.26803911],
        [ 0.27528917]],

       [[ 0.24245064],
        [ 0.26846557],
        [ 0.24202418],
        ...,
        [-0.06290496],
        [-0.08081688],
        [-0.06333144]],

       ...,

       [[ 0.01951179],
        [ 0.01093518],
        [ 0.03109021],
        ...,
        [-0.02980372],
        [ 0.0491011 ],
        [-0.23607121]],

       [[-0.21205669],
        [-0.30597058],
        [-0.34199232],
        ...,
        [ 0.09970309],
        [ 0.07182911],
        [ 0.10785088]],

       [[ 0.10484906],
        [ 0.1039914 ],
        [ 0.10570673],
        ...,
        [ 0.00621804],
        [ 0.01436582],
        [ 0.02551542]]])

The following corresponds to reshaping the existing 3d data to 2d and converting it to .csv. The csv file was created on google drive and later downloaded to the local pc

In [ ]:
pp_x_rshp=pp_x.reshape(189560,3000)

In [ ]:
pp_x_rshp

array([[ 0.03913197,  0.08437123,  0.10246694, ..., -0.04998939,
        -0.05360853, -0.01017883],
       [-0.05179896, -0.05496571, -0.02148865, ...,  0.28387638,
         0.23320841,  0.25899479],
       [ 0.29609098,  0.18661197,  0.23908951, ...,  0.16218277,
         0.16580191,  0.1535873 ],
       ...,
       [-0.03455384,  0.00189911,  0.05185315, ..., -0.0525553 ,
         0.025301  ,  0.01360005],
       [-0.00620155,  0.02170071, -0.06515632, ...,  0.0352018 ,
         0.0203506 ,  0.0437525 ],
       [ 0.12205883, -0.08855822, -0.07505712, ..., -0.06470629,
        -0.08810818, -0.10835982]])

In [ ]:
pp_X_sub_482_rshp=pp_X_sub_482.reshape(3070,3000)

In [ ]:
pp_x_rshp_df=pd.DataFrame(pp_x_rshp)

In [ ]:
pp_x_rshp_df

0         1         2     ...      2997      2998      2999
0       0.039132  0.084371  0.102467  ... -0.049989 -0.053609 -0.010179
1      -0.051799 -0.054966 -0.021489  ...  0.283876  0.233208  0.258995
2       0.296091  0.186612  0.239090  ...  0.162183  0.165802  0.153587
3       0.134134  0.116491  0.116491  ...  0.126896  0.205612  0.274829
4       0.333640  0.359878  0.381141  ...  0.057680  0.122372  0.207422
...          ...       ...       ...  ...       ...       ...       ...
189555  0.381730  0.488388  0.302523  ... -0.263172  0.131060 -0.558396
189556 -0.620501 -0.653354 -0.647503  ... -0.020153  0.041502  0.006850
189557 -0.034554  0.001899  0.051853  ... -0.052555  0.025301  0.013600
189558 -0.006202  0.021701 -0.065156  ...  0.035202  0.020351  0.043752
189559  0.122059 -0.088558 -0.075057  ... -0.064706 -0.088108 -0.108360

[189560 rows x 3000 columns]

In [ ]:
pp_x_rshp_df['sleep_stage']=y

In [ ]:
pp_x_rshp_df

0         1         2  ...      2998      2999  sleep_stage
0       0.039132  0.084371  0.102467  ... -0.053609 -0.010179            0
1      -0.051799 -0.054966 -0.021489  ...  0.233208  0.258995            0
2       0.296091  0.186612  0.239090  ...  0.165802  0.153587            0
3       0.134134  0.116491  0.116491  ...  0.205612  0.274829            0
4       0.333640  0.359878  0.381141  ...  0.122372  0.207422            0
...          ...       ...       ...  ...       ...       ...          ...
189555  0.381730  0.488388  0.302523  ...  0.131060 -0.558396            0
189556 -0.620501 -0.653354 -0.647503  ...  0.041502  0.006850            0
189557 -0.034554  0.001899  0.051853  ...  0.025301  0.013600            0
189558 -0.006202  0.021701 -0.065156  ...  0.020351  0.043752            0
189559  0.122059 -0.088558 -0.075057  ... -0.088108 -0.108360            0

[189560 rows x 3001 columns]

In [ ]:
pp_X_sub_482_rshp_df=pd.DataFrame(pp_X_sub_482_rshp)

In [ ]:
pp_X_sub_482_rshp_df

0         1         2     ...      2997      2998      2999
0    -0.236906 -0.234348 -0.218142  ...  0.008743 -0.012155  0.018125
1     0.085082  0.021110  0.123464  ...  0.251833  0.268039  0.275289
2     0.242451  0.268466  0.242024  ... -0.062905 -0.080817 -0.063331
3    -0.134126 -0.107258 -0.096170  ... -0.081670 -0.126023 -0.032199
4     0.009596  0.047125  0.021963  ...  0.048405 -0.019405 -0.001066
...        ...       ...       ...  ...       ...       ...       ...
3065  0.237787  0.278525  0.310259  ... -0.387877 -0.398598 -0.360432
3066 -0.405459 -0.413607 -0.363863  ...  0.086409  0.078262  0.086838
3067  0.019512  0.010935  0.031090  ... -0.029804  0.049101 -0.236071
3068 -0.212057 -0.305971 -0.341992  ...  0.099703  0.071829  0.107851
3069  0.104849  0.103991  0.105707  ...  0.006218  0.014366  0.025515

[3070 rows x 3000 columns]

In [ ]:
pp_X_sub_482_rshp_df['sleep_stage']=y_sub_482

In [ ]:
pp_X_sub_482_rshp_df

0         1         2  ...      2998      2999  sleep_stage
0    -0.236906 -0.234348 -0.218142  ... -0.012155  0.018125            0
1     0.085082  0.021110  0.123464  ...  0.268039  0.275289            0
2     0.242451  0.268466  0.242024  ... -0.080817 -0.063331            0
3    -0.134126 -0.107258 -0.096170  ... -0.126023 -0.032199            0
4     0.009596  0.047125  0.021963  ... -0.019405 -0.001066            0
...        ...       ...       ...  ...       ...       ...          ...
3065  0.237787  0.278525  0.310259  ... -0.398598 -0.360432            0
3066 -0.405459 -0.413607 -0.363863  ...  0.078262  0.086838            0
3067  0.019512  0.010935  0.031090  ...  0.049101 -0.236071            0
3068 -0.212057 -0.305971 -0.341992  ...  0.071829  0.107851            0
3069  0.104849  0.103991  0.105707  ...  0.014366  0.025515            0

[3070 rows x 3001 columns]

In [ ]:
pp_x_rshp_df.to_csv(r'/gdrive/My Drive/filtered_seventy_four_subs.csv', header=False, index=False)

In [ ]:
pp_X_sub_482_rshp_df.to_csv(r'/gdrive/My Drive/filtered_x_sub_482.csv', header=False, index=False)

In [ ]:
Fs=100
#batch_size=10

30 epochs

In [ ]:
checkpoint_path_erfan_model = "/gdrive/My Drive/physionet/dt_shuf2_phy/sub_482_75th_sub_with_50_sub_tr_dt/weights-phy_shuf2_30_epochs.ckpt"

In [ ]:
checkpoint_path_erfan_model_dir = os.path.dirname(checkpoint_path_erfan_model)

In [ ]:
cp_callback_erfan_model_30_epochs = tf.keras.callbacks.ModelCheckpoint(
    filepath=checkpoint_path_erfan_model,
    monitor='accuracy', 
    verbose=1, 
    save_weights_only=True,
    mode = 'max',
    period=30)

In [ ]:
x_shuf2

0         1         2     ...      2997      2998      2999
30619  0.028727  0.019679 -0.133682  ...  0.112420 -0.050894 -0.031441
66149 -0.064884 -0.093265 -0.132826  ... -0.018443 -0.026183 -0.034784
48811 -0.059500  0.146904 -0.240929  ...  0.162455  0.134180  0.155858
92761 -0.000707 -0.031338 -0.033694  ... -0.025211 -0.019085 -0.030866
55338  0.041534  0.042803  0.060566  ... -0.004143 -0.009642  0.008545
...         ...       ...       ...  ...       ...       ...       ...
51593  0.031014  0.023804  0.028469  ... -0.032604 -0.033452 -0.022849
18057 -0.384249 -0.389748 -0.375541  ... -0.245389 -0.275636 -0.285259
85379 -0.008700 -0.039090 -0.050882  ...  0.008989 -0.005525 -0.000990
26462  0.011455  0.039213  0.040054  ... -0.066774 -0.046586 -0.033127
68917  0.024599  0.027144  0.011875  ...  0.035626  0.047077  0.064042

[116718 rows x 3000 columns]

In [ ]:
model_30_epochs = model_b(verbose=VBS)
hist_30_epochs = model_30_epochs.fit(
    x_shuf2_rshp, y_shuf2_, batch_size=batch_size, epochs=30, verbose=VBS, callbacks=[cp_callback_erfan_model_30_epochs]
)

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 inLayer (InputLayer)           [(None, 3000, 1)]    0           []                               
                                                                                                  
 fConv1 (Conv1D)                (None, 500, 64)      3264        ['inLayer[0][0]']                
                                                                                                  
 cConv1 (Conv1D)                (None, 60, 32)       12832       ['inLayer[0][0]']                
                                                                                                  
 fMaxP1 (MaxPooling1D)          (None, 62, 64)       0           ['fConv1[0][0]']                 
                                                                                              

In [ ]:
y_pred_shuf1 = model_30_epochs.predict(pp_X_sub_482, batch_size=batch_size)
y_pred_shuf1 = np.array([np.argmax(s) for s in y_pred_shuf1])
f1_cnn_shuf1 = f1_score(y_sub_482, y_pred_shuf1, average="macro")
if VBS:
    print("F1 score: {}".format(f1_cnn_shuf1))
    report_shuf1 = classification_report(y_sub_482, y_pred_shuf1)
    print(report_shuf1)

F1 score: 0.7200930070849723
              precision    recall  f1-score   support

           0       0.98      0.95      0.96      1351
           1       0.32      0.19      0.23       187
           2       0.72      0.96      0.82       807
           3       0.95      0.72      0.82       318
           4       0.85      0.68      0.76       407

    accuracy                           0.85      3070
   macro avg       0.76      0.70      0.72      3070
weighted avg       0.85      0.85      0.84      3070



30 epochs

F1 score: 0.735178728137819
              precision    recall  f1-score   support

           0       0.97      0.96      0.97      1351
           1       0.42      0.17      0.24       187
           2       0.74      0.95      0.83       807
           3       0.93      0.78      0.85       318
           4       0.84      0.74      0.79       407

    accuracy                           0.86      3070
    macro avg       0.78      0.72      0.74      3070
    weighted avg       0.86      0.86      0.85      3070


   



  
   


# New Section

50 subjects

30 epochs: 86%

In [ ]:
latest_30_epochs=tf.train.latest_checkpoint(checkpoint_path_erfan_model_dir)

In [ ]:
#print(latest_30_epochs)
epochs_30_model_path="/content/gdrive/My Drive/physionet/dt_shuf2_phy/sub_482_75th_sub_with_50_sub_tr_dt/weights-phy_shuf2_30_epochs.ckpt"

In [ ]:
loaded_model_30_epochs = model_b(verbose=VBS)
loaded_model_30_epochs.load_weights(epochs_30_model_path)

Model: "model_2"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 inLayer (InputLayer)           [(None, 3000, 1)]    0           []                               
                                                                                                  
 fConv1 (Conv1D)                (None, 500, 64)      3264        ['inLayer[0][0]']                
                                                                                                  
 cConv1 (Conv1D)                (None, 60, 32)       12832       ['inLayer[0][0]']                
                                                                                                  
 fMaxP1 (MaxPooling1D)          (None, 62, 64)       0           ['fConv1[0][0]']                 
                                                                                            

In [ ]:
optimizer = Adam(lr=1e-4)

In [ ]:
loaded_model_30_epochs.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])

In [ ]:
y_sub_482_=np_utils.to_categorical(y_sub_482)

In [ ]:
y_sub_482_

array([[1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       ...,
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.],
       [1., 0., 0., 0., 0.]], dtype=float32)

In [ ]:
accuracy_loaded_model_30_epochs = loaded_model_30_epochs.evaluate(pp_X_sub_482, y_sub_482_)

96/96 [==============================] - 2s 5ms/step - loss: 0.4161 - accuracy: 0.8616
